### PR TITLE
[GSK-2280] Feature: Added Number-to-Words Transformation

### PR DESCRIPTION
**Pull request by sagar118**
## Description

I have added the functionality to perform number-to-words transformation using the `num2words` library. I have added the unit test cases and dependency to the `pdm.lock` file.

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

## Checklist

- [x] I've read the [`CODE_OF_CONDUCT.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I've read the [`CONTRIBUTING.md`](https://github.com/Giskard-AI/ai-inspector/blob/master/CONTRIBUTING.md) guide.
- [ ] I've updated the code style using `make codestyle`.
- [x] I've written tests for all new methods and classes that I created.
- [ ] I've written the docstring in Google format for all the methods and classes that I used.

### CODE REVIEW

**Review by mattbit — CHANGES_REQUESTED** [review]
Thanks a lot for this _very high-quality_ PR!

I added a few comments, but the main improvement over this would be to support multiple languages in the same dataset, without imposing a given language at the transformation level. We have metadata on the `Dataset` object that can be used to retrieve the detected language → you can check how this is done in the `TextGenderTransformation`.

Don’t hesitate to ping me if something is not clear, it’s a tricky part of the code. Also don’t hesitate to refactor other classes if you feel it’s necessary.

File: giskard/scanner/robustness/text_transformations.py [hunk 1]
```diff
@@ -147,6 +148,22 @@ def make_perturbation(self, text):
         return "".join(pieces)
 
 
+class TextNumberToWordTransformation(TextTransformation):
+    name = "Transform numbers to words"
+
```
[the remark this answers is not in the document]
Comment (sagar118) [reply giskard/scanner/robustness/text_transformations.py:153]: Thank you for the tip, I'll look into it.

File: tests/scan/test_text_transformations.py [hunk 2]
```diff
@@ -115,6 +115,77 @@ def test_punctuation_strip_transformation():
...
+        transformed_text[5]
+        == "For more information, call us at +one-eight hundred-five hundred and fifty-five-one hundred and ninety-nine."
+    )
+
+    # German tests
+    t = TextNumberToWordTransformation(column="text", lang="de")
+
+    transformed = dataset.transform(t)
+    transformed_text = transformed.df.text.values[6:8]
+    assert transformed_text[0] == "Der Zug soll um neun:fünfundvierzig Uhr ankommen."
```
[the remark this answers is not in the document]
Comment (sagar118) [reply tests/scan/test_text_transformations.py:169]: I added those example for the reason to showcase the different interactions with the num2words library. I believe there is no solution directly available to format these cases.

File: tests/scan/test_text_transformations.py [hunk 3]
```diff
@@ -115,6 +115,77 @@ def test_punctuation_strip_transformation():
...
+                "You can access the report at https://www.example123.com/report2023.",  # Don't transform URLs
+                "The serial number of the device is XC1234-AB56. Please send your queries to contact123@service4u.com.",  # Don't transform emails
+                "The dataset contains approximately 1500 entries spanning 5 years.",
+                "The total cost of the project is estimated to be around $4,500.75.",
+                "For more information, call us at +1-800-555-0199.",
+                "Der Zug soll um 09:45 Uhr ankommen.",  # German
+                "Die Gesamtkosten der Artikel betragen 45,67 Dollar.",  # German
+                "La tasa de éxito de este procedimiento es del 98,6%.",  # Spanish
+                "Hoy corrió 13,2 millas en el maratón.",  # Spanish
+                "Son anniversaire est le 22.08 du mois.",  # French
```
[the remark this answers is not in the document]
Comment (sagar118) [reply tests/scan/test_text_transformations.py:132]: I agree, it was a typo from my end. I'll make the changes necessary.

File: tests/scan/test_text_transformations.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sagar118) [reply tests/scan/test_text_transformations.py:169]: I would like to know your thoughts on these cases, should we use the results the num2words library provides, or want to format the text for particular edge cases like date/time format and numbers in different languages?

[the remark this answers is not in the document]
Comment (mattbit) [reply tests/scan/test_text_transformations.py:169]: It just to understand if this we could get some improvement without too much work. Since it seems not possible, let’s keep those examples which are very good for this test as they show how the edge cases are dealt with.

[the remark this answers is not in the document]
Comment (sagar118) [reply tests/scan/test_text_transformations.py:169]: Thank you for your input. I'll look into how we can make some improvements for these edge cases and retain those examples in the test function.

**Comment by mattbit** [comment]
Hey @sagar118, unless you’ve already started looking into this, we can have someone from the team to take over and do the final changes :)

**Comment by sagar118** [comment]
Hey @mattbit, I apologize for the delay. I have worked on it a bit. If it's okay with you I can continue working on it and get back to you ASAP. Is there a timeline you're looking for to wrap up this issue?

File: giskard/scanner/robustness/text_transformations.py [hunk 4]
```diff
@@ -209,6 +210,21 @@ def _switch(self, word, language):
             return None
 
 
+class TextNumberToWordTransformation(TextLanguageBasedTransformation):
+    name = "Transform numbers to words"
+
+    def __init__(self, column):
+        super().__init__(column)
```
Comment (mattbit) [inline giskard/scanner/robustness/text_transformations.py:217]: we don't need this, right?

File: pyproject.toml [hunk 5]
```diff
@@ -21,12 +21,6 @@ giskard = [
 ]
 
 
-[[tool.pdm.source]]
-type = "find_links"
-url = "https://download.pytorch.org/whl/cpu/torch_stable.html"
-name = "torch"
-
-
```
Comment (mattbit) [inline pyproject.toml:29]: ?

File: tests/llm/test_llm_client.py
Comment (mattbit) [inline tests/llm/test_llm_client.py:1]: this does not belong here?

**Review by mattbit — COMMENTED** [review]
Some extraneous code that maybe you should check @kevinmessiaen (problems with the automatic merge?)

**Comment by kevinmessiaen** [comment]
> Some extraneous code that maybe you should check @kevinmessiaen (problems with the automatic merge?)

Fixed this, it was a merge issue indeed

**Review by kevinmessiaen — APPROVED** [review]
